### PR TITLE
init agent implementation + test

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
ref #504 

heres a draft pr for some of the ideas in the ref issue. going to add more tests before publishing + im having some issue getting standard to recognize WeakRef and FinalizationRegistry

### CODE REVIEW

File: lib/agent.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    constructor (opts) {
+        this[kAgentPoolOpts] = opts
+        this[kAgentCache] = new Map()
+        this[kAgentRegistry] = new FinalizationRegistry(key => {
+            // get the Pool WeakRef from the cache
+            const ref = this[kAgentCache].get(key)
+            // if the WeakRef exists and the object has been reclaimed
+            // according to mdn the second logic operation here is superfluous as a deref call in a FR will always return undefined
+            if (ref !== undefined && ref.deref() === undefined) {
+                // remove the WeakRef from the cache
```
Comment (ronag) [inline lib/agent.js:18]: Would be nice with an assertion checking that there are not pending requests in the pool

File: lib/agent.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                this[kAgentCache].delete(key)
+            }
+        })
+    }
+
+    get (origin) {
```
Comment (ronag) [inline lib/agent.js:24]: check origin is non empty string

File: lib/agent.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+const globalAgent = new Agent({
+    connections: 16,
+    pipelining: 2,
+    headersTimeout: 10e3,
+    bodyTimeout: 10e3
+})
+
+function request (url, opts = {}) {
+    url = url instanceof URL ? url : new URL(url)
```
Comment (ronag) [inline lib/agent.js:55]: would be nice if we support url-like POJO

File: lib/agent.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    const agent = opts.agent == undefined ? globalAgent : opts.agent
+
+    const pool = agent.get(url.origin)
+    return pool.request({
+        ...opts,
+        method: opts.method || 'GET',
+        path: url.path || `${url.pathname || ''}${url.search || ''}`
+    })
+}
+
```
Comment (ronag) [inline lib/agent.js:66]: Should also export the 2 (3) other API methods

File: lib/agent.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        return pool
+    }
+}
+
+const globalAgent = new Agent({
+    connections: 16,
+    pipelining: 2,
+    headersTimeout: 10e3,
+    bodyTimeout: 10e3
+})
```
Comment (ronag) [inline lib/agent.js:52]: Should it be possible to override the global agent?

File: lib/agent.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        // add the Pool to the finalization registry
+        this[kAgentRegistry].register(pool, origin)
+
+        return pool
+    }
+}
+
+const globalAgent = new Agent({
+    connections: 16,
+    pipelining: 2,
```
Comment (ronag) [inline lib/agent.js:49]: I think we need to go without pipelining for default

File: lib/agent.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    headersTimeout: 10e3,
+    bodyTimeout: 10e3
```
Comment (ronag) [inline lib/agent.js:51]: Use default timeouts?

File: lib/agent.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            // according to mdn the second logic operation here is superfluous as a deref call in a FR will always return undefined
+            if (ref !== undefined && ref.deref() === undefined) {
+                // remove the WeakRef from the cache
+                this[kAgentCache].delete(key)
+            }
+        })
+    }
+
+    get (origin) {
+        // check the cache for an existing Pool
```
Comment (ronag) [inline lib/agent.js:25]: Later we might need to support 'origin' as an object containing more info to make the key from, in particular TLS stuff.

File: lib/agent.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    headersTimeout: 10e3,
+    bodyTimeout: 10e3
+})
+
+function request (url, opts = {}) {
+    url = url instanceof URL ? url : new URL(url)
+
+    const agent = opts.agent == undefined ? globalAgent : opts.agent
+
+    const pool = agent.get(url.origin)
```
Comment (ronag) [inline lib/agent.js:59]: since agent can be implemented in user land we should probably check the return value here.

File: lib/agent.js (hunk 1 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:18]: what should this callback do in the case there are requests pending or in flight? Also any ideas how to test this scenario? As far as I understand it the GC would not clean up a Pool currently being used

Comment (ronag) [reply lib/agent.js:18]: > what should this callback do in the case there are requests pending or in flight? 

It shouldn't be possible. If that happens it's a bug which is why I suggest an assertion.

Comment (ronag) [reply lib/agent.js:18]: Actually never mind. It's not possible to assert this.

File: lib/agent.js (hunk 4 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:66]: should this match the existing http.Agent ?

File: lib/agent.js (hunk 5 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:52]: What about:

```js
let globalAgent = new Agent()

function setGlobalAgent(agent) {
  if (agent instanceof Agent) {
    globalAgent = agent
  } else {
    throw TypeError('agent must be instance of Agent')
  }
}

function request () { ... }
```

File: lib/agent.js (hunk 6 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:49]: ill just make the default globalAgent `new Agent()` and rely on defaulting within the constructor

File: lib/agent.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      if (ref !== undefined && ref.deref() === undefined) {
+        // remove the WeakRef from the cache
+        this[kAgentCache].delete(key)
+      }
+    })
+  }
+
+  get (origin) {
+    if (typeof origin !== 'string' || origin === '') {
+      throw TypeError('Origin must be a non-empty string.')
```
Comment (ronag) [inline lib/agent.js:27]: use an error from `errors.js`.

File: lib/agent.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ *
+ * @param {string | import('url').URL} url
+ * @param {*} opts
+ */
+function request (url, opts = {}) {
+  // if url is a string transform into URL inst. otherwise use as is
+  url = typeof url === 'string' ? new URL(url) : url
+
```
Comment (ronag) [inline lib/agent.js:62]: ensure origin and path (or pathname) are valid strings.

File: lib/agent.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  const agent = (opts.agent === undefined || opts.agent === null) ? globalAgent : opts.agent
```
Comment (ronag) [inline lib/agent.js:63]: `opts.agent == null`

File: lib/agent.js (hunk 4 above)
Comment (ronag) [reply lib/agent.js:77]: http.Agent? I mean stream and pipeline.

File: lib/agent.js [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+  const pool = agent.get(url.origin)
```
Comment (ronag) [inline lib/agent.js:65]: agent could return a regular client as well... no need to assume it's a Pool. Just an object with `Client.request` trait

File: lib/agent.js [hunk 14]
```diff
@@ -0,0 +1,81 @@
+'use strict'
+/* global WeakRef, FinalizationRegistry */
+
+const { Pool } = require('../')
+
+const kAgentPoolOpts = Symbol('agent pool opts')
+const kAgentCache = Symbol('agent cache')
+const kAgentRegistry = Symbol('agent registry')
```
Comment (ronag) [inline lib/agent.js:8]: missing new line

File: lib/agent.js [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+class Agent {
+  constructor (opts) {
+    this[kAgentPoolOpts] = opts
+    this[kAgentCache] = new Map()
+    this[kAgentRegistry] = new FinalizationRegistry(key => {
+      // get the Pool WeakRef from the cache
+      const ref = this[kAgentCache].get(key)
+      // if the WeakRef exists and the object has been reclaimed
+      // according to mdn the second logic operation here is superfluous as a deref call in a FR will always return undefined
```
Comment (ronag) [inline lib/agent.js:17]: I actually think the second call is necessary as there is a possible ref race otherwise. See the Chrome page on WeakRefs.

Comment (ronag) [reply lib/agent.js:17]: https://github.com/tc39/proposal-weakrefs

File: lib/agent.js (hunk 12 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:63]: standard doesn't like that so I had to be more declarative

Comment (ronag) [reply lib/agent.js:63]: ah, intersting

File: lib/agent.js (hunk 11 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:62]: For validation:
origin **cannot** be empty
path and pathname **can** be empty

Comment (ronag) [reply lib/agent.js:62]: I think you only need to validate origin. path is validated by core.

File: lib/agent.js (hunk 15 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:17]: Ooh I also realized me putting all of this in the class is wrong too. V8 page on WeakRefs is very in depth

File: lib/agent.js (hunk 4 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:77]: Ahh I see what you mean. Adding two more functions `stream` and `pipeline` that rely on the Agent being created here.

File: lib/agent.js [hunk 16]
```diff
@@ -1,56 +1,68 @@
...
-    const pool = new Pool(origin, this[kAgentPoolOpts])
-    // add a WeakRef of the Pool to the cache
-    this[kAgentCache].set(origin, new WeakRef(pool))
-    // add the Pool to the finalization registry
-    this[kAgentRegistry].register(pool, origin)
+let globalAgent = new Agent()
 
-    return pool
+function setGlobalAgent (agent) {
+  if (agent instanceof Agent) {
```
Comment (ronag) [inline lib/agent.js:59]: Doesn't need to be an instance of Agent. Just implement a get method.

File: lib/agent.js [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+    // otherwise, if it isn't in the cache or the reference has been cleaned up, create a new one!
+    const value = fn(key)
+    // add a WeakRef of the value to the cache
+    cache.set(key, new WeakRef(value))
+    // add the value to the finalization registry
+    cleanup.register(value, key)
+    return value
+  }
+}
```
Comment (ronag) [inline lib/agent.js:38]: Move to util?

Comment (Ethan-Arrowood) [reply lib/agent.js:38]: this is agnostic enough to have its own module but for now i'll move it to util

File: test/agent.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
+const tap = require('tap')
+const http = require('http')
+const { request } = require('../lib/agent')
+
+tap.test('request a resource using globalAgent', t => {
```
Comment (ronag) [inline test/agent.js:5]: Only enable for Node 14+

File: lib/agent.js [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+    method: opts.method || 'GET',
+    path: url.path || `${url.pathname || ''}${url.search || ''}`
+  })
+}
+
+module.exports = {
+  request,
+  setGlobalAgent,
+  Agent
+}
```
Comment (ronag) [inline lib/agent.js:98]: Should we export these in index.js?

File: test/agent.js (hunk 18 above)
Comment (Ethan-Arrowood) [reply test/agent.js:5]: was going to ask about this... what should I do in the actual implementation if they aren't using Node 14+? Load a polyfill dep?

File: lib/agent.js (hunk 19 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:98]: I think so! Have something along the lines of:
```js
const undici = require('undici')

const agent = new undici.Agent()

const result = await undici.request('https://example.com', { agent })
```

File: test/agent.js (hunk 18 above)
Comment (ronag) [reply test/agent.js:5]: Just don't export it?

File: lib/agent.js [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+    throw InvalidArgumentError('Argument agent must be instance of Agent')
+  }
+}
+
+/**
+ *
+ * @param {string | import('url').URL} url
+ * @param {*} opts
+ */
+function request (url, opts = {}) {
```
Comment (ronag) [inline lib/agent.js:71]:
```suggestion
function request (url, { agent = globalAgent, ...opts } = {}) {
```

File: test/agent.js (hunk 18 above)
Comment (Ethan-Arrowood) [reply test/agent.js:5]: This is going to be impossible to type. I propose we throw an error instead?

Comment (Ethan-Arrowood) [reply test/agent.js:5]: Lol i'm copy/pasting the util i just wrote in case we need it later:
```js
function nodeMajorVersionIsGreaterThanOrEqualTo ({ major: expectedMajor, minor: expectedMinor, patch: expectedPatch }) {
  const [ foundMajor, foundMinor, foundPatch ] = process.versions.node.split('.').map(str => parseInt(str, 10))

  return foundMajor >= expectedMajor && foundMinor >= expectedMinor && foundPatch >= expectedPatch
}
```

**Comment by Ethan-Arrowood** [comment]
Last commit has a first attempt at adding the other two methods. its probably wrong but i'll test and make it better tomorrow

**Review by ronag — COMMENTED** [review]
LGTM, but needs more tests

File: lib/agent.js [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  return client
+}
+
+/**
+ *
+ * @param {string | import('url').URL} url
+ * @param {*} opts
+ */
```
Comment (ronag) [inline lib/agent.js:66]: Either add this to everything in file or remove it.

Comment (Ethan-Arrowood) [reply lib/agent.js:66]: im going to remove it for now - and soon ill revisit doing this for the whole project

**Comment by ronag** [comment]
@dnlup

File: lib/agent.js [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+function getClientFromAgent (agent, url) {
+  if (!nodeMajorVersionIsGreaterThanOrEqualTo(14)) {
+    throw NotSupportedError('Node.js version must be 14 or above')
+  }
+
+  // if url is a string transform into URL inst. otherwise use as is
+  url = typeof url === 'string' ? new URL(url) : url
+  // ensure it atleast has a non-empty string origin
+  if (typeof url.origin !== 'string' || url.origin === '') {
```
Comment (ronag) [inline lib/agent.js:49]: url can be nully here

**Comment by Ethan-Arrowood** [comment]
Any thoughts on this part:

```js
{
  path: url.path || `${url.pathname || ''}${url.search || ''}`
}
```

Should we just stick to url.path?

File: index.js [hunk 23]
```diff
@@ -3,6 +3,7 @@
 const Client = require('./lib/core/client')
 const errors = require('./lib/core/errors')
 const Pool = require('./lib/pool')
+const { Agent, request, stream, pipeline, setGlobalAgent } = require('./lib/agent')
```
Comment (Ethan-Arrowood) [inline index.js:6]: @ronag so this is causing a circular dependency issue. 

`lib/agent.js` requires `Pool` from this file since it needs the prototype methods on lines 14-18. How would you like me to resolve this? Can I move the prototype changes to `lib/pool.js`?

Comment (ronag) [reply index.js]: What prototype methods? I don't see how this is circular?

File: lib/agent.js [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+const { InvalidArgumentError, InvalidReturnValueError, NotSupportedError } = require('./core/errors')
+const { Pool } = require('../')
+const { weakCache, nodeMajorVersionIsGreaterThanOrEqualTo } = require('./core/util')
+
+const kAgentFactory = Symbol('agent factory')
+
+class Agent {
+  constructor (opts) {
+    /* istanbul ignore next */
+    if (!nodeMajorVersionIsGreaterThanOrEqualTo(14)) {
```
Comment (dnlup) [inline lib/agent.js:12]: I did some checks and `FinalizationRegistry` seems supported from version `14.6.0`. Wouldn't be better to check if `FinalizationRegistry` and `WeakRef` are defined globally? Assuming users might use version `14.5.0` or something like that.

File: lib/agent.js [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+    }
+    this[kAgentFactory] = weakCache(origin => new Pool(origin, opts))
+  }
+
+  get (origin) {
+    if (typeof origin !== 'string' || origin === '') {
+      throw new InvalidArgumentError('Origin must be a non-empty string.')
+    }
+
+    return this[kAgentFactory](origin)
```
Comment (dnlup) [inline lib/agent.js:23]: Should we support unix socket also?

**Comment by dnlup** [comment]
> Any thoughts on this part:
> 
> ```js
> {
>   path: url.path || `${url.pathname || ''}${url.search || ''}`
> }
> ```
> 
> Should we just stick to url.path?

That should cover all cases, what doubts do you have about it?

File: lib/agent.js (hunk 25 above)
Comment (ronag) [reply lib/agent.js:23]: nice to have

File: lib/agent.js (hunk 24 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:12]: yeah good point - i saw how you did it in your PR and will copy that pattern

File: index.js (hunk 23 above)
Comment (Ethan-Arrowood) [reply index.js]: lines 14-18 are like `Pool.prototype.request = require('client-request.js')`

Agent requires access to this modded Pool class rather than the one exported from `./lib/pool`. 

So since `agent.js` requires `../index.js` and `index.js` requires `./lib/agent` its a circular ref

File: lib/agent.js [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+
+  if (typeof agent.get === 'function') {
+    globalAgent = agent
+  } else {
+    throw new InvalidArgumentError('Argument agent must implement Agent')
+  }
+}
+
+function getPath (url) {
+  return url.path || `${url.pathname || ''}${url.search || ''}`
```
Comment (dnlup) [inline lib/agent.js:43]: On second thought, I think `pathname` should be `||` with `'/'`. If `url` is an instance of  `URL` the right side will never be reached (`pathname` always defaults to `'/'`), otherwise we should reproduce the same behavior (i.e., the user passes a plain object)

**Comment by Ethan-Arrowood** [comment]
> That should cover all cases, what doubts do you have about it? 

Just haven't seen this before. I'll move it to the util file and test it

File: lib/agent.js (hunk 26 above)
Comment (dnlup) [reply lib/agent.js:43]: We should also include `hash`:

```bash
URL {
  href: 'http://localhost/?test=1#readme',
  origin: 'http://localhost',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost',
  hostname: 'localhost',
  port: '',
  pathname: '/',
  search: '?test=1',
  searchParams: URLSearchParams { 'test' => '1' },
  hash: '#readme'
}
```

**Comment by Ethan-Arrowood** [comment]
I have to be missing something obvious - it seems like the CI is blatantly ignoring the SKIP I added to the test and still running the Agent stuff 😕

File: lib/agent.js (hunk 26 above)
Comment (Ethan-Arrowood) [reply lib/agent.js]: Is there some existing utility method for this kind of pathing? I feel weird adding this support when it feels like theres lots of edge cases

**Review by mcollina — APPROVED** [review]
lgtm

File: index.js (hunk 23 above)
Comment (ronag) [reply index.js:6]: Implement pool in agent with require('./lib/client-request') etc...

Comment (ronag) [reply index.js:6]: alterantively move the pool modding to the pool file

File: lib/agent.js (hunk 26 above)
Comment (dnlup) [reply lib/agent.js]: I don't know, unfortunately. I think this

```js
`${url.pathname || ''}${url.search || ''}${url.hash || ''}`
```

given how `hash` and `search` are parsed:

```bash
> new URL('http://localhost/?test=1#readme')
URL {
  href: 'http://localhost/?test=1#readme',
  origin: 'http://localhost',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost',
  hostname: 'localhost',
  port: '',
  pathname: '/',
  search: '?test=1',
  searchParams: URLSearchParams { 'test' => '1' },
  hash: '#readme'
}
```
```bash
> new URL('http://localhost/#readme?test=1')
URL {
  href: 'http://localhost/#readme?test=1',
  origin: 'http://localhost',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost',
  hostname: 'localhost',
  port: '',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {},
  hash: '#readme?test=1'
}
```

Do you see other corner cases?

File: lib/agent.js (hunk 25 above)
Comment (dnlup) [reply lib/agent.js:22]: The `socketPath` option should be used in `origin` to identify a pool of connections. Also, it should be passed to the `kAgentFactory`. It can be done in a new PR if it's not urgent.

File: lib/agent.js [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+  return (url, { agent = globalAgent, method = 'GET', ...opts } = {}, ...additionalArgs) => {
+    /* istanbul ignore next */
+    checkForWeakRefAndFinalizationRegistrySupport()
+
+    if (url === undefined || url === null) {
+      throw new InvalidArgumentError('Argument url must be defined')
+    }
+
+    // if url is a string transform into URL inst. otherwise use as is
+    url = typeof url === 'string' ? new URL(url) : url
```
Comment (dnlup) [inline lib/agent.js:57]: The last thing I am bringing up because I have seen it brought up in the past, using `URL` here is acceptable in terms of performance drop?

File: lib/agent.js (hunk 10 above)
Comment (dnlup) [reply lib/agent.js]: I think this one is resolved 😄

**Review by dnlup — APPROVED** [review]
None of the comments I made are blocking, lgtm

File: lib/agent.js (hunk 25 above)
Comment (Ethan-Arrowood) [reply lib/agent.js:22]: I'm +1 for follow up PR as this one is already quite large

File: lib/agent.js (hunk 26 above)
Comment (Ethan-Arrowood) [reply lib/agent.js]: @ronag @mcollina thoughts on this?

**Review by mcollina — APPROVED** [review]
lgtm

**Review by ronag — CHANGES_REQUESTED** [review]
LGTM, need docs

**Comment by ronag** [comment]
thanks a lot!

File: lib/agent.js (hunk 27 above)
Comment (ronag) [reply lib/agent.js:58]: This api is more for convenience than performance so I don't see it as a huge issue. But yea, there is room for optimization.
